### DISCUSSION TITLE
메시지를 받을 수 있는 시점

### DISCUSSION THREAD

**Discussion by chaejin-jen** [42MVP/ft_transcendence#2] [General]
메시지를 받을 수 있는 시점 논의

1️⃣  온라인 상태
2️⃣  클라이언트가 채팅룸 뷰를 보고 있는 상태
3️⃣  채널은 채팅룸 뷰, DM 은 온라인일 때

====================================
ps. 오프라인 상태에서 메시지 받기는 메시지를 DB에 저장 안하니까 불가능 (socket 쪽에 인메모리로 메시지를 못 받은 유저가 접속할 때 까지 계속 들고있다면 가능)

**Comment by chaejin-jen** [comment]
- 2️⃣ 라 생각하고 소켓 명세 작성함
- 2️⃣ -> 제일 편해보임, 3️⃣ -> DM 은 오프라인일 때도 되야 된다 생각하는데, 불가능하니 그나마 이거라도

**Comment by khk811** [comment]
이 문제에 대해 방향성을 정하기 전에 추상적인 레벨을 올려서 전체적으로 어떻게 동작하는지를 확실히 알아야 할것 같습니다.

그러니까
1. API던 소켓 이벤트던 클라이언트에서 채팅방에 입장하거나 입력신호가 갈꺼고,
2. 그걸 서버가 무엇이던 어떤 방식으로 받아서 해당 입력에 대한 처리를 할껍니다.
3. 그러면 서버가 전송한 emit 내지는 예외 등등의 신호를 클라이언트가 받고,
4. 받은 내용을 브라우저에 반영을 한다.

이 이슈에서는 4에서 받는 내용에 대해 휘발을 어느단계 까지 할 것인지 결정을 해야 한다고 말씀해 주셨어요.

슬랙에서 선택받은 `3️⃣ 채널은 채팅룸 뷰, DM 은 온라인일 때`를 구현한다고 하면,
채널에 작성한 메시지는 채팅룸 뷰를 볼 때까지만 존재하기 때문에 
```
1. 채팅룸 끄면 -> 게임소켓 채팅소켓 따로해야함, 보내도 온 거 모름 (못보내게 예외처리 할지 말지 정해야함)
2. 새로고침 전까지 -> 프론트에서 전역 변수로 들고 있어야 함
3. 브라우저 캐시 지우기 전까지 -> 프론트에서 저장
4. 다른 브라우저 가도 -> DB 또는 백엔드 인메모리로
```
위와 같은 케이스에서 채팅룸 뷰는 1번에 해당되고,
DM은 로그아웃 전까지 존재해야 하기 때문에 2번에 해당되는게 맞나요?

**Comment by chaejin-jen** [comment]
현재 논의 된 휘발 시점 : `탭 또는 창을 닫을 때`
- 메시지 저장은 `session storage` 에 하기로 함
- local storage 에 저장 할 경우, 다른 유저로 로그인 한 경우, 이전 유저의 대화 내역이 뜨기 때문


휘발 단계 부분 다시 정정할게요! 저 땐, socket.io 에서 room 의 존재도 몰랐고, session storage 도 몰라서 저렇게 써놨습니다.

```
1. 채팅룸 끄면 -> 채팅 뷰 보고 있을 동안만 (실시간 채팅만 가능함)
2. 새로고침 전까지 -> 프론트에서 전역 변수로 들고 있어야 함
3. 브라우저 탭이나 창 닫기 전에 -> 프론트에서 session storage로 저장
4. 다른 브라우저 가도 -> DB 또는 백엔드 인메모리로
```

A. `메시지 받을 수 있는 시점`과 `채팅 휘발 시점`은 별개 입니다. 

메시지 받을 수 있는 시점은 room에 join 을 언제해주냐이고, `채널뷰를 볼 때` 해줄꺼냐, `로그인 된 시점`에서 해줄거냐 입니다. 
메시지 휘발 시점은 백에서 저장 안하기로 했으니까, 프론트에서 저장할지 말지의 문제입니다.
현재, session storage 에 저장하기로 했으니까, 채널이나 DM이 동일하게 창을 닫기 전까지 유효합니다.


2️⃣ 의 경우 chats 뷰 에서 `채팅룸이 열릴 때`(보일 때)  백엔드 소켓에서 해당 room 을 join 하고, `다른 뷰로 넘어갈 때`, room 을 `leave` 함.
3️⃣ 의 경우, DM 은 `로그인 시점`에서 생성되어 있는 DM(room) 에 모두 join 해주고, channel 은 채널뷰에서 해당 채널 열었을 때 해당 room 에 join 해 줌. 


저는 이렇게 생각하고 있었는데, 같은 생각을 한건지 다른 분들도 의견주시면 좋겠습니다 !

**Reply by sinryuji** [reply]
제 생각엔 메세지를 받는 시점이나 휘발 시점을 채널이든 DM이든 구별을 안두는 쪽으로 가는게 좋아 보입니다!!

**Reply by khk811** [reply]
만약 채팅방 기준으로 `채팅뷰를 본 시점`에서만 메시지를 받을 수 있다고 하면, room join이 API 호출 안에서 이루어져야 할것 같습니다.
채팅 룸 리스트 가져오는 API를 호출하는 시점이 `매번 채팅 뷰를 볼때`인지 `최초에 한번`만 까지 결정을 한다면,
형키님이 위에 말씀해 주신것처럼 
- `메시지를 받을수 있는 시점`을 채널과 DM을 동일하게 `2️⃣ 클라이언트가 채팅룸 뷰를 보고 있는 상태`로 진행하고
- `매번 채팅 뷰`를 볼 때 `채팅 룸 리스트 API`를 호출하고, 그때 그 API 내에서 `room join`을 한다

이러면 어떨까요?

**Comment by chaejin-jen** [comment]
결론 4️⃣  온라인일 때 !!!!

로그인 시점에서 기존에 들어가있던 채팅방에 다 들어가짐 (소켓 룸에 조인)

socket 이벤트핸들러 join
- ❌ 실시간 채팅 가능한 유저
- ❌ socket room join
- ⭕ 백이 클라이언트에게 보냄
- ⭕ 새로운 유저가 채널에 입장시 기존 유저들의 유저리스트를 업데이트하기 위함


채널 리스트 가져오는 API 요청 시점
- 채팅 뷰 켰을 때 (처음 켯을 때만 해도 될 듯)


기타의견

- hyeongki : 강퇴 동작은 REST API 에서 하는게 좋을 것 같다 생각함
